WooCommerce/Classes/ViewRelated/Custom Fields/CustomFieldsListViewModel.swift

```diff
@@ -1,29 +1,49 @@
 import Combine
 import Foundation
+import Networking
+import Yosemite
 
 final class CustomFieldsListViewModel: ObservableObject {
-    private let originalCustomFields: [CustomFieldViewModel]
-
-    var shouldShowErrorState: Bool {
-        savingError != nil
-    }
+    private let stores: StoresManager
+    private var originalCustomFields: [CustomFieldViewModel]
+    private let customFieldsType: MetaDataType
+    private let siteId: Int64
+    private let parentItemId: Int64
 
     @Published var selectedCustomField: CustomFieldUI? = nil
     @Published var isAddingNewField: Bool = false
+    @Published var isSavingChanges: Bool = false
 
-    @Published private(set) var savingError: Error?
     @Published private(set) var combinedList: [CustomFieldUI] = []
     @Published var notice: Notice?
 
-    @Published private var editedFields: [CustomFieldUI] = []
-    @Published private var addedFields: [CustomFieldUI] = []
-    @Published private var deletedFieldIds: [Int64] = []
+    @Published private var pendingChanges = PendingCustomFieldsChanges()
+    private var editedFields: [CustomFieldUI] {
+        get { pendingChanges.editedFields }
+        set { pendingChanges = pendingChanges.copy(editedFields: newValue) }
+    }
+    private var addedFields: [CustomFieldUI] {
+        get { pendingChanges.addedFields }
+        set { pendingChanges = pendingChanges.copy(addedFields: newValue) }
+    }
+    private var deletedFieldIds: [Int64] {
+        get { pendingChanges.deletedFieldIds }
+        set { pendingChanges = pendingChanges.copy(deletedFieldIds: newValue) }
+    }
```

Comment:
Personally I think the current approach is less error prone, with it we have a single source of truth that is `pendingChanges`, the computed properties are just a syntactic sugar, while if we have the other fields as backed properties as well, then we have different source of truths that we should keep in sync using the `combineLatest` that you suggested.
So unless this is a blocker, I'll keep my approach 🙂

Reply:
Sorry, I don't quite understand your explanation. The observation of the properites using Combine works just the same as your get-set, and in both cases `pendingChanges` are still the only source of truth. And in both cases, if you have a new property for `pendingChanges`, you have to update to code to consider any new changes from that property.

The only thing I see is that my proposal makes it shorter and easier to follow. The less spread out the logic, the less chance there is to accidentally remove something breaking the code.

Please let me know if there's something I'm missing here.

Reply:
To showcase the issue, please test the following patch (on top of this branch) and run the test `test_showcase_multiple_source_of_truths_issue`

<details>
<summary> Patch </summary>

```patch
diff --git a/WooCommerce/Classes/ViewRelated/Custom Fields/CustomFieldsListViewModel.swift b/WooCommerce/Classes/ViewRelated/Custom Fields/CustomFieldsListViewModel.swift
index 7572e775c3..ad4a55c2a6 100644
--- a/WooCommerce/Classes/ViewRelated/Custom Fields/CustomFieldsListViewModel.swift	
+++ b/WooCommerce/Classes/ViewRelated/Custom Fields/CustomFieldsListViewModel.swift	
@@ -17,19 +17,11 @@ final class CustomFieldsListViewModel: ObservableObject {
     @Published private(set) var combinedList: [CustomFieldUI] = []
     @Published var notice: Notice?
 
-    @Published private var pendingChanges = PendingCustomFieldsChanges()
-    private var editedFields: [CustomFieldUI] {
-        get { pendingChanges.editedFields }
-        set { pendingChanges = pendingChanges.copy(editedFields: newValue) }
-    }
-    private var addedFields: [CustomFieldUI] {
-        get { pendingChanges.addedFields }
-        set { pendingChanges = pendingChanges.copy(addedFields: newValue) }
-    }
-    private var deletedFieldIds: [Int64] {
-        get { pendingChanges.deletedFieldIds }
-        set { pendingChanges = pendingChanges.copy(deletedFieldIds: newValue) }
-    }
+    @Published private(set) var pendingChanges = PendingCustomFieldsChanges()
+    @Published private(set) var editedFields: [CustomFieldUI] = []
+    @Published private var addedFields: [CustomFieldUI] = []
+    @Published private var deletedFieldIds: [Int64] = []
+
     @Published private(set) var hasChanges: Bool = false
 
     init(customFields: [CustomFieldViewModel],
@@ -123,6 +115,12 @@ extension CustomFieldsListViewModel {
 
         isSavingChanges = false
     }
+
+    func buggyCode() {
+        editedFields = [CustomFieldUI(key: "key", value: "value", fieldId: 1)]
+
+        pendingChanges = PendingCustomFieldsChanges()
+    }
 }
 
 private extension CustomFieldsListViewModel {
@@ -173,6 +171,10 @@ private extension CustomFieldsListViewModel {
         $pendingChanges
             .map { $0.hasChanges }
             .assign(to: &$hasChanges)
+
+        $editedFields.combineLatest($addedFields, $deletedFieldIds)
+            .map { PendingCustomFieldsChanges(editedFields: $0, addedFields: $1, deletedFieldIds: $2) }
+            .assign(to: &$pendingChanges)
     }
 
     @MainActor
diff --git a/WooCommerce/WooCommerceTests/ViewRelated/Custom Fields/CustomFieldsListViewModelTests.swift b/WooCommerce/WooCommerceTests/ViewRelated/Custom Fields/CustomFieldsListViewModelTests.swift
index f052e3898f..c8e143c0c5 100644
--- a/WooCommerce/WooCommerceTests/ViewRelated/Custom Fields/CustomFieldsListViewModelTests.swift	
+++ b/WooCommerce/WooCommerceTests/ViewRelated/Custom Fields/CustomFieldsListViewModelTests.swift	
@@ -253,4 +253,10 @@ final class CustomFieldsListViewModelTests: XCTestCase {
         // Then: An error should be thrown
         XCTAssertNotNil(viewModel.notice)
     }
+
+    func test_showcase_multiple_source_of_truths_issue() {
+        viewModel.buggyCode()
+
+        XCTAssertEqual(viewModel.pendingChanges.editedFields.count, viewModel.editedFields.count)
+    }
 }
```
</details>

The patch above showcases that we have different source of truths using your solution, the `combineLatest` tries to hide this by syncing the changes from the three individual properties to the wrapping class, but if someone edited the `pendingChanges` property directly, they would break this syncing.
While my solution even though it's more verbose, it relies on a single source of truth (`pendingChanges`), the other properties don't have a storage of their own, and when modified they always forward changes to the same source of truth.
Please let me know if I'm missing something in my example above, I'm fine with applying your solution is you still feel it's better 🙂.

Reply:
Thanks, now I understand the complication comes from having two places for `editedFields`, `addedFields`, and `deletedFieldIds`. It would be nicer if we could keep them in `pendingChanges` only, and make the view model update `pendingChanges` when it needs to update any of those fields. 

I don't want to be a blocker and make you make that change, so please feel free to keep your implementation. Next time, I hope we can simplify the code by not duplicating the same properties in two places like this 😅.

Reply:
I can delete them if you prefer so, but for me (and as I said before), they are a syntactic sugar that helps in making the code that consumes them easy to follow (even with the added verbosity in the definition of the properties), and I think this is one of the good uses of computed properties (in both Kotlin and Swift).